### DISCUSSION TITLE
readQuery Performance

### DISCUSSION THREAD

**Discussion by thdxr** [urql-graphql/urql#2579] [Q&A] [answered]
I'm seeing a simple `readyQuery` take 1-2ms to resolve.
The query itself is simple and should basically be a quick lookup by type + id
I know the cache isn't exposed to avoid writing directly to it but is there a way to directly read from it? I'd expect this to be 100x faster

**Comment by kitten** [comment, marked as answer]
Ok, look 😅 there's just been an issue here, but I'll repeat some points here to summarise https://github.com/FormidableLabs/urql/issues/2548

- don't measure in a non-prod build
- you can't just say that you expect it to be n times faster, that's unfortunately not how this works
- performance depends a lot on the traversal complexity, so an absolute number of Xms won't cut it

I'd love to tell you I could make it order of magnitudes faster, but it's literally as far as I can tell at the limit of how fast Graphcache can be given the current constraints. You may or may not also see significant changes depending on the query complexity, size of the payload, production builds, other GC pressure in your code, etc 😅

**Reply by thdxr** [reply]
This is a simple "get object by id" reading directly from the cache is actually 500-1000x faster.  I didn't make that up, I tested but to be fair it's non-production so let me try it in prod.

Is there no chance of providing an escape hatch to access the cache directly? I did this through things I probably should not be doing

**Reply by thdxr** [reply]
Tested in prod and no difference.

To clarify my situation is not exactly the same as the issue you linked.  While I do have a lot of entries in the cache - I am only pulling out **one**.  If I hack into the cache and get it directly from the `base` map this is unsurprisingly fast.

I want to understand a bit better what the overhead is when going through `readQuery`.  If I cannot work around that then I'll have to build a specialized data layer for this which I'd love to avoid if possible.

**Reply by JoviDeCroock** [reply]
Mind showing the query? Also, any optimistic layers/... going on?

**Reply by thdxr** [reply]
The query is like this
```
query {
  record {
    id
    name
  }
}
```

No optimistic layers, it hits a `resolver` which returns `{ __typename: "record", id: args.id }` and then gets sent to the cache I imagine

**Reply by JoviDeCroock** [reply]
So when you look at the trace where is most time being spent? It's hard to think why this would be so much slower apart from maybe you need to hoist the `gql` invocation etc but generally it should be fast, i.e. graphcache is already significantly faster compared to others 😅 you could also try `cache.resolve` to see if that makes a diff

**Reply by thdxr** [reply]
I think maybe I might have measured in the wrong place, I'm noticing the UI lag only when the offline storage is turned on.  Does the entire offline storage not get read into memory on start?

Also I would like to read from the cache but it's not accessible outside of the resolvers

**Reply by JoviDeCroock** [reply]
It is hydrated upon first load yes, I mean we are moving to a different topic atm 😅 we have explained a lot of the time why we limit cache-access to GraphQL operations. We have generally not found any cases where we wouldn't be opening up a lot of pitfalls to our users. https://github.com/FormidableLabs/urql/pull/2076#issuecomment-1047333942

**Reply by thdxr** [reply]
OK I tracked this down further, I made a shim for the storage and I'm seeing that `writeData` is being called whenever I do a `cache.readQuery` which is what is slowing everything down. It's being passed an empty object

Any ideas?

**Reply by JoviDeCroock** [reply]
We always call it for every operation it looks like [here](https://github.com/FormidableLabs/urql/blob/28f11dc4220528a1d9cfceb3b2725054664cb810/exchanges/graphcache/src/store/data.ts#L186) I would have to look deeper in as to why we call it with an empty object but generally this is expected to be called for every operation as a successful query, new state after mutation, ... needs to be written

**Reply by thdxr** [reply]
Yeah I wouldn't expect a cache read to trigger this. And I guess since my state is big it's slow to write - which seems to imply it's saving the whole state every time as opposed to a granular update?

Either way for now I did a quick check to see if the object is empty and skipped the write, solves my issue

**Reply by kitten** [reply]
No, it's saving a granular update, so that shouldn't be a problem. The problem is kind of the measurement tbh. IndexedDB can be extremely slow depending on the browser, and React, especially in development, thrashes the stack aggressively and it all becomes a lot of work that's hard to differentiate. So, the next non-IE11 version is likely going to tweak this timing 

But the DB updates should never be 100x slower. Anyway, that also kind of depends on whether the measurements are accurate though, without accurate numbers everything's more like an educated guess

**Reply by thdxr** [reply]
I'm not using React. You can ignore the measurements entirely, I can see when I turn storage on visually everything gets laggy both in development and production.  And when I insert my shim to skip writes for empty objects, it speeds up. I mainly am asking why a cache readQuery triggers a write (empty or otherwise) to IndexedDB

I can visually see with my eyes that the app gets laggy when I don't have my patch in there. I understand you might be getting questions all day from people making incorrect assumptions and that wears down your patience but it is a bit frustrating your reaction to everything I say is to find a way to dismiss it.

**Reply by kitten** [reply]
My patience has little to do with this. I also don't mind people making wrong assumptions. Honestly though, the one thing in Open Source that takes the most time is triaging and investigating.

For context, even given this problem we'll likely have to reproduce at least part of it, investigate why it's slow, address other things along the way if we have time, then fix it. Even if the fix is a one liner, even if the minimum we'll do is add the check whether anything has to be written before calling the write method. There's more attached to this with timing, IndexedDB having a higher cost here than anticipated, which means if it really does take long, we'll have to act on that, and so on and so forth.

So, when I'm trying to provide you more information on what's going on, when we're asking questions because we're getting too few information, when we're trying to fit the pieces together, then it's not a "project-wide patience problem" but it's a time budgeting problem in this very thread.

I'd also love to donate time to this without people having to have assumptions made about my patience or why I phrased a response a certain way. I simply spend little time on them, if I think little time has been spent on the preceding message.

To get back to the topic, this is clearly why I addressed this; you're not using React, and you're saying you can see the slowdown, where originally you said it takes 1-2ms, which is clearly an order of magnitude of a difference at least.

Also, the measurement isn't unnecessary, so I'd still love to know more about them. That's because beyond the fix that will prevent this work unnecessarily, we'll ultimately now have to investigate of what the overhead of IDB storage is when it isn't unnecessary.
